modules/crm_core_match/src/Plugin/crm_core_match/field/Name.php

```diff
@@ -2,41 +2,96 @@
+   * A query factory.
    *
-   * The name field uses text and select fields to set values we will need to pass information into other field handlers
-   * to get the right records to pass back.
-   *
-   * @see DefaultMatchingEngineFieldType::fieldRender()
+   * @var \Drupal\Core\Entity\Query\QueryFactory
+   */
+  protected $queryFactory;
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    $field = $configuration['field'];
+    unset($configuration['field']);
+    $name = new static(
+      $field,
+      $container->get('entity.query')->get('crm_core_contact', 'AND'),
+      $configuration,
+      $plugin_id,
+      $plugin_definition
+    );
+    $name->queryFactory = $container->get('entity.query');
+    return $name;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getOperators($property = 'value') {
+    return array(
+      'CONTAINS' => t('Contains'),
+    );
+  }
+
+  /**
+   * {@inheritdoc}
    */
-  public function fieldRender($field, $field_info, &$form) {
-    foreach ($field_info['columns'] as $item => $info) {
-      if ($field['settings']['inline_css'][$item] != 'display:none') {
-        $field_item['field_name'] = $field['field_name'];
-        $field_item['label'] = $field['label'] . ': ' . $field_info['settings']['labels'][$item];
-        $field_item['bundle'] = $field['bundle'];
-        $field_item['field_item'] = $item;
-
-        if (isset($field['settings'][$item . '_field']) && $field['settings'][$item . '_field'] == 'select') {
-          $item = new selectMatchField();
-          $item->fieldRender($field_item, $field_info, $form);
+  public function match(ContactInterface $contact, $property = 'value') {
+    $field = $this->field->getName();
+    $name = $contact->get($field)->{$property};
+    $parts = preg_split('/[\ \,]+/', $name);
+    $valid_parts = [];
+    foreach ($parts as $part) {
+      if (strlen($part) > 2) {
+        $valid_parts[] = $part;
+      }
+    }
+    $result = [];
+    if (!empty($valid_parts)) {
+      foreach ($valid_parts as $part) {
+        $this->query = $this->queryFactory->get('crm_core_contact', 'AND');
+        $this->query->condition('type', $contact->bundle());
+        if ($contact->id()) {
+          $this->query->condition('contact_id', $contact->id(), '<>');
         }
-        else {
-          $item = new textMatchField();
-          $item->fieldRender($field_item, $field_info, $form);
+
+        if ($field instanceof FieldConfigInterface) {
+          $field .= '.' . $property;
+        }
+
+        $this->query->condition($field, $part, 'CONTAINS');
+        $ids = $this->query->execute();
+        foreach ($ids as $id) {
+          if (isset($result[$id])) {
+            $result[$id] += 1;
+          }
+          else {
+            $result[$id] = 1;
+          }
         }
       }
     }
+    arsort($result);
+    return $result;
```

Comment:
As discussed, the resulting score should be proportional to the configured boost score (`$this->getScore()`).